pkg/clusteragent/autoscaling/workload/controller_horizontal.go

```diff
@@ -126,70 +150,295 @@ func (hr *horizontalController) computeScaleAction(
 		}
 	}
 
 	// No specific policy defined, defaulting to allow
 	return true, ""
 }
+
+func applyScaleUpPolicy(
+	currentTime time.Time,
+	events []datadoghq.DatadogPodAutoscalerHorizontalAction,
+	policy *datadoghq.DatadogPodAutoscalerScalingPolicy,
+	currentDesiredReplicas, targetDesiredReplicas int32,
+) (int32, time.Duration, string) {
+	if policy == nil {
+		return targetDesiredReplicas, 0, ""
+	}
+
+	strategy := datadoghq.DatadogPodAutoscalerMaxChangeStrategySelect
+	// If no strategy is defined, we default to the max policy for scale up
+	if policy.Strategy != nil {
+		strategy = *policy.Strategy
+	}
+
+	var maxReplicasFromRules int32
+	var selectStrategyFunc func(int32, int32) int32
+	minExpireIn := time.Hour // We don't support more than 1 hour of events
+	if strategy == datadoghq.DatadogPodAutoscalerMinChangeStrategySelect {
+		maxReplicasFromRules = math.MaxInt32
+		selectStrategyFunc = min
+	} else {
+		maxReplicasFromRules = math.MinInt32
+		selectStrategyFunc = max
+	}
+
+	for _, rule := range policy.Rules {
+		// We could find directly `periodStartReplicas` by looking at `FromReplicas` in the first matching event.
+		// TODO: In case of manual scaling (outside of DPA), we could consider it in the calculation, while it's currently not.
+		replicasAdded, replicasRemoved, numEvents, expireIn := accumulateReplicasChange(currentTime, events, rule.PeriodSeconds)
+		minExpireIn = min(minExpireIn, expireIn)
+		if numEvents == 0 && rule.Match != nil && *rule.Match == datadoghq.DatadogPodAutoscalerIfScalingEventRuleMatch {
+			// Rule should be skipped as no scaling events were found in the window
+			continue
+		}
+
+		periodStartReplicas := currentDesiredReplicas - replicasAdded + replicasRemoved
+		var ruleMax int32
+		if rule.Type == datadoghq.DatadogPodAutoscalerPodsScalingRuleType {
+			ruleMax = periodStartReplicas + rule.Value
+		} else if rule.Type == datadoghq.DatadogPodAutoscalerPercentScalingRuleType {
+			// 1.x * start may yield the same number of replicas as periodStartReplicas, ceiling up to always always allow at least 1 replica
+			// otherwise it would block scaling up forever.
+			ruleMax = int32(math.Ceil(float64(periodStartReplicas) * (1 + float64(rule.Value)/100)))
+		}
+		maxReplicasFromRules = selectStrategyFunc(maxReplicasFromRules, ruleMax)
+	}
+
+	// No rules matched, not restricting the scaling
+	if maxReplicasFromRules == math.MaxInt32 || maxReplicasFromRules == math.MinInt32 {
+		return targetDesiredReplicas, 0, ""
+	}
+
+	// If we already above what we are allowed to scale to, we should not scale further
+	if currentDesiredReplicas > maxReplicasFromRules {
+		maxReplicasFromRules = currentDesiredReplicas
+	}
+
+	// If we're limited by rules,
+	if targetDesiredReplicas > maxReplicasFromRules {
+		return maxReplicasFromRules, minExpireIn, fmt.Sprintf("desired replica count limited to %d (originally %d) due to scaling policy", maxReplicasFromRules, targetDesiredReplicas)
+	}
+	return targetDesiredReplicas, 0, ""
```

Comment:
Why do we return `0` here ? Shouldn't we reevaluate after `minExpireIn` in case the next event limits us ?